src/ucp/dt/datatype_iter.inl

```diff
@@ -530,10 +572,29 @@ ucp_datatype_iter_is_end(const ucp_datatype_iter_t *dt_iter)
 static UCS_F_ALWAYS_INLINE void
 ucp_datatype_memh_dereg(ucp_context_h context, ucp_mem_h *memh_p)
 {
-    if (*memh_p != NULL) {
-        ucp_memh_put(context, *memh_p);
-        *memh_p = NULL;
+    if (*memh_p == NULL) {
+        return;
+    }
+
+    /* Check if memh is external (user memh), and if so, don't touch it */
+    if (ucp_memh_is_user_memh(*memh_p)) {
+        return;
     }
+
+    /* If it's internal, release it */
+    ucp_memh_put(context, *memh_p);
+    *memh_p = NULL;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_datatype_iter_check_memh_mds(ucp_context_h context, ucp_mem_h memh,
```

Comment:
there's a possibility that we'll remove this registration, because seems there's no simple way to do it.
Waiting for Yossi's response here:
https://github.com/openucx/ucx/pull/8514#discussion_r983741153